cli/azd/docs/environment-with-hooks.md

```diff
@@ -0,0 +1,172 @@
+sleep 5
+```
+
+By using `declare -A azdEnv`, you can allocate a set to write the keys and values from the `azd env get-values` output. Then you can access the values by referencing the environment keys with `"${azdEnv["ENV_VAR_FROM_AZD"]}"`. This strategy would allow you to test or run your script either manually or using `azd hooks run <hook name>`, and without affecting the terminal variables. 
+
+Here is the powershell equivalent.
+
+```pwsh
+$azdEnv = @{}
+$azdEnvRaw = azd env get-values
+
+foreach ($line in $azdEnvRaw -split "`n") {
+    if ($line -match '^(.*?)=(.*)$') {
+        $key = $matches[1]
+        $value = $matches[2] -replace '^"|"$'
+        $azdEnv[$key] = $value
+    }
+}
+
+Write-Output $azdEnv["ENV_VAR_FROM_AZD"]
+
+# just to let you see the output
+Start-Sleep -Seconds 5
+```
+
+This strategy is fine and easy to implement when you have only one script, but if you have many scripts, you would need to copy and paste the same code in all the scripts, which might be hard to maintain. For such scenarios, take a look to the next strategy.
+
+#### Use terminal scoped variables
+
+When you have multiple scripts and you need all of some of them to access azd environment, you can make your script to set the azd environment in your terminal variables. You have to be careful with this strategy and ideally make sure to restore your terminal variables once all your scripts have run. This is because some terminals like powershell does not create one environment for executing the script, but update the environment of the terminal you are running in.
+
+The main benefit of using your terminal variables is that you just need to set the variables one time, at your starting point, and then all scripts you run after will read the azd environment as direct variables from the terminal. See the next example:
+
+```bash
+#!/bin/bash
+
+declare -A azdEnv
+while IFS='=' read -r key value; do
+    value=$(echo "$value" | sed 's/^"//' | sed 's/"$//')
+    export "$key=$value"
+done <<EOF
+$(azd env get-values)
+EOF
+
+echo $ENV_VAR_FROM_AZD
+
+# just to let you see the output
+sleep 5
+```
+
+And the powershell version:
+
+```pwsh
+foreach ($line in (& azd env get-values)) {
+    if ($line -match "([^=]+)=(.*)") {
+        $key = $matches[1]
+        $value = $matches[2] -replace '^"|"$'
+	    [Environment]::SetEnvironmentVariable($key, $value)
+    }
+}
+
+Write-Output $env:ENV_VAR_FROM_AZD
+
+# just to let you see the output
+Start-Sleep -Seconds 5
+```
+
+After running your script, make sure that azd environment is not leaked and persisted to your terminal variables, as it might affect future execution of azd commands because azd won't be switching to any other environment (unless you force it with the -e flag). This is because the `AZURE_ENV_NAME` would be persisted as terminal variable. So, depending on the terminal you are using (for sure on powershell), you would need to have a `restore variables` script at the end.
```

Comment:
Maybe provide an example of this script for completeness?
Mine are @ https://github.com/Azure-Samples/openai-chat-app-entra-auth-builtin/blob/main/scripts/unload_azd_env.ps1  and https://github.com/Azure-Samples/openai-chat-app-entra-auth-builtin/blob/main/scripts/unload_azd_env.sh
but Copilot wrote the Powershell and I havent tested it yet.

Reply:
I was going to, but, there is some risk on the `unload_azd_env.ps1`.  Imagine you have an existing variable in your system, like the AZURE_LOCATION (because that's what you always want to use).  After running unload, your variable will be removed.

So, what we need is not just unload, but to capture the initial variables state and restore it at the end.  I was going to show that pattern but it got night and wanted to have some sleep, lol.   I will add it soon.  Thank you for looking at this

Reply:
Ah okay yep great, makes sense! Will follow your lead.